### PR TITLE
Add PureComponent fallback babel transform

### PR DESCRIPTION
**Pull request by majapw**
Fix for https://github.com/airbnb/react-dates/issues/1354

Never got around to doing this before and now it is blocking.

Thanks @skevy for writing the transform!

to: @ljharb @skevy @Sumeet-Jain

### CODE REVIEW

File: .babelrc [hunk 1]
```diff
@@ -7,7 +7,8 @@
           "svgo": false
         }],
         ["transform-replace-object-assign", { "moduleSpecifier": "object.assign" }],
-        "istanbul"
+        "istanbul",
```
Comment (ljharb) [inline .babelrc]: i think the istanbul transform needs to come last, so it can properly instrument the code for coverage

**Comment by majapw** [comment]
@ljharb @skevy i think this makes sense now!

File: scripts/pure-component-fallback.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
+module.exports = function pureComponentFallback(babel) {
+  const { types: t, template } = babel;
```
Comment (ljharb) [inline scripts/pure-component-fallback.js]: nbd but this can be destructured in the previous line

File: scripts/pure-component-fallback.js [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+      t.blockStatement([template('return shallowCompare(this, nextProps, nextState);')()]),
+    );
+    return method;
+  };
+
+  function superclassIsPureComponent(path) {
+    const superclass = path.get('superClass');
+    // check for PureComponent
+    if (t.isIdentifier(superclass)) {
+      return superclass.node.name === 'PureComponent';
```
Comment (ljharb) [inline scripts/pure-component-fallback.js]: probably fine for "just this repo", but if we'll want to use this in other projects, we should make sure that `PureComponent` is actually a named import from `'react'` or a destructure from React imported/required from `'react'`

File: scripts/pure-component-fallback.js [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    if (t.isMemberExpression(superclass)) {
+      // Check for React.PureComponent
+      return superclass.get('object').node.name === 'React'
```
Comment (ljharb) [inline scripts/pure-component-fallback.js]: same here

File: src/components/CalendarDay.jsx [hunk 5]
```diff
@@ -340,4 +338,4 @@ export default withStyles(({ reactDates: { color, font } }) => ({
   CalendarDay__today: {},
   CalendarDay__firstDayOfWeek: {},
   CalendarDay__lastDayOfWeek: {},
-}), { pureComponent: pureComponentAvailable })(CalendarDay);
+}), { pureComponent: typeof React.PureComponent !== 'undefined' })(CalendarDay);
```
Comment (ljharb) [inline src/components/CalendarDay.jsx]: not a blocker, but it'd be nicer if the transform could add this instead

File: scripts/pure-component-fallback.js [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  const buildPureOrNormalSuperclass = () => t.LogicalExpression('||',
+    t.memberExpression(t.identifier('React'), t.identifier('PureComponent')),
+    t.memberExpression(t.identifier('React'), t.identifier('Component')));
+
+  const buildShouldComponentUpdate = () => {
+    const method = t.classMethod(
+      'method',
+      t.identifier('shouldComponentUpdate'),
+      [t.identifier('nextProps'), t.identifier('nextState')],
+      t.blockStatement([template('return shallowCompare(this, nextProps, nextState);')()]),
```
Comment (ljharb) [inline scripts/pure-component-fallback.js]: the trailing function comma here is breaking node 4.

**Pull request merged**